scripts/diff-workflow-against-master.sh

```diff
@@ -0,0 +1,471 @@
+	cache_dir="$parent/cache"
+	diffs_dir="$parent/diffs"
+	logs_dir="$parent/logs"
+	declare -gr assets_dir
+	declare -gr cache_dir
+	declare -gr diffs_dir
+	declare -gr logs_dir
+	echo "Initializing storage area: $parent"
+
+	# Don't trust content from a prior interrupted run
+	if [[ -f "$parent/.interrupted" ]]; then
+		rm -rf "$parent"
+	fi
+
+	# Make the directories if they don't exist
+	for dir in "$assets_dir" "$cache_dir" "$diffs_dir" "$logs_dir"; do
+		if [[ ! -d "$dir" ]]; then
+			mkdir -p "$dir"
+		# Otherwise, purge content older than 5-minutes
+		else
+			for filename in "$dir"/*; do
+				if [[ "$(seconds_old "$filename")" -gt 300 ]]; then
+					rm -rf "$filename"
+				fi
+			done
+		fi
+	done
+	# If the user Ctrl-C'd the job, then some files might be
+	# partially written, so drop a breadcrumb to clean up next run.
+	# (we could just blow away the content here, but we want to
+	# let the user inspect content after interrupting the run.)
+	trap 'touch $parent/.interrupted' INT
+}
+
+##
+#  Ensures all pre-requisites are setup and have passed
+#  before we start making REST queries and writing files.
+#
+function init() {
+	init_baseurl
+	init_local_branch
+	init_netrc
+	init_dirs
+}
+
+##
+#  Downloads a file if we otherwise don't have it.
+#  (Note that the script on launch cleans up files older than
+#  5 minutes, so most of the time we'll be downloading.)
+#
+function download() {
+	local url="$1"
+	local outfile="$2"
+	if [[ ! -f "$outfile" ]]; then
+		curl --silent     \
+		     --location   \
+		     --netrc      \
+		     "$url"       \
+		     -o "$outfile"
+	fi
+}
+
+##
+#  Unzips files inside their containing directory.
+#  Clobbers existing files.
+#
+function unpack() {
+	local zipfile="$1"
+	local zipdir
+	zipdir="$(dirname "$zipfile")"
+	pushd "$zipdir" > /dev/null
+	unzip -qq -o "$zipfile"
+	rm -f "$zipfile"
+	popd > /dev/null
+}
+
+##
+#  Constructs and fetches REST urls using our personal access
+#  token. Files are additionally hashed based on the REST URL
+#  and cached.  This allows for rapid-rerunning without needing
+#  to hit GitHub's API again (for duplicate requests). This
+#  avoid us exceeding our repo limit on API calls/day.
+#
+function pull() {
+	# Buildup the REST URL by appending arguments
+	local url="$baseurl"
+	for element in "$@"; do
+		url="$url/$element"
+	done
+	local url_hash
+	url_hash="$(echo "$url" | md5sum | cut -f1 -d' ')"
+	local outfile="${cache_dir}/${url_hash}.json"
+	if [[ ! -f "$outfile" ]]; then
+		download "$url" "$outfile"
+	fi
+	cat "$outfile"
+}
+
+##
+#  Gets one or more keys from all records
+#
+function get_all() {
+	local container="$1"
+	local return_keys="$2"
+	jq -r '.'"$container"'[] | '"${return_keys}"
+}
+
+##
+#  Gets one or more return_key(s) from records that have
+#  matching search_key and search_value hits
+#
+function query() {
+	local container="$1"
+	local search_key="$2"
+	local search_value="$3"
+	local return_keys="$4"
+	jq -r --arg value "$search_value"\
```

Comment:
`./scripts/diff-workflow-against-master.sh: line 274: jq: command not found`

Add a dependency check on script start; if a user has no `jq` installed, the script should immediately exit with an error (before a user has a chance to generating a token).

Reply:
I know that token is saved before this happens, but some user might get lost and generate a new one.

Reply:
Huge catch here; I'm baffled why I didn't do a requirements check first - thank you @dreamer !
Will run a check for `unzip` and any other non-core-util commands I might be using.

Reply:
Added (for all tools)